### PR TITLE
CORE-13770. Improve sxs support in ldr

### PR DESCRIPTION
**Pull request by yagoulas**
Right now our ldr doesn't support loading two different versions of libraries like comctl32 when they are installed in winsxs. This PR aims to fix this.

### CODE REVIEW

**Comment by yagoulas** [comment]
It should be noted that this PR needs A LOT of testing.

File: dll/ntdll/rtl/libsupp.c [hunk 1]
```diff
@@ -642,6 +642,139 @@ RtlPcToFileHeader(IN PVOID PcValue,
     return ImageBase;
 }
 
+/* NOTE: Remove this one once our actctx support becomes better */
+NTSTATUS find_actctx_dll( PUNICODE_STRING pnameW, LPWSTR *fullname, PUNICODE_STRING CallerBuffer, PUNICODE_STRING DynamicString)
```
Comment (learn-more) [inline dll/ntdll/rtl/libsupp.c]: Why is DynamicString a PUNICODE_STRING in this function? it's only checked to be a valid pointer.

File: dll/ntdll/ldr/ldrutils.c [hunk 2]
```diff
@@ -336,13 +236,105 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
+                    {
+                        if (ShowSnaps)
+                        {
+                            DPRINT1("LDR: %Z got redirected to %wZ\n", &ImportNameAnsi, RedirectedImportName);
+                        }
+                        /* Let Ldrp know */
+                        RedirectedDll = TRUE;
+                    }
+
+                    if (RedirectedDll || Status == STATUS_SXS_KEY_NOT_FOUND)
```
Comment (paulsapps) [inline dll/ntdll/ldr/ldrutils.c]: This whole block of logic is repeated at least twice, could be refactored in to a helper?

**Comment by jeditobe** [comment]
Is it CORE-13770 ?

File: boot/bootdata/packages/reactos.dff.in [hunk 3]
```diff
@@ -70,6 +70,10 @@ Signature = "$ReactOS$"
 51 = system32\CatRoot
 52 = system32\CatRoot2
 53 = AppPatch
+54 = winsxs
+55 = winsxs\manifests
+56 = winsxs\x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.2600.2982_none_deadbeef
+57 = winsxs\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.2600.2982_none_deadbeef
```
[the remark this answers is not in the document]
Comment (wjk) [reply boot/bootdata/packages/reactos.dff.in]: @learn-more A quick spot-check on a Windows 10 system I have access to reveals that the equivalent manifest files are named as such:

```
amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.10586.0_none_396e892957c7fb25.manifest
amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.10586.0_none_8c15ae12515e1c22.manifest
amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.10586.839_none_a2ddba30a539a2ae.manifest
x86_microsoft.windows.common-controls_6595b64144ccf1df_5.82.10586.0_none_811bc0006c44242b.manifest
x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.10586.0_none_d3c2e4e965da4528.manifest
x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.10586.839_none_ea8af107b9b5cbb4.manifest
```

Thus, the first major and minor parts of the version numbers always stay the same, yet the build and revision parts can and do change to match the build and revision numbers of the version of Windows they were shipped in. (In this case, the computer is running the November update of Windows 10, build 10586. The files ending in `10586.0` are the original copies of these assemblies that were shipped in that release; the files ending in `10586.839` are patched versions delivered via Windows Update.)

That said, this should not cause _that_ much of a problem, because the SxS manifest code I use in my apps lists the version as `6.0.0.0`. Windows then looks for and loads a version that is _at least_ `6.0.0.0`, even if the version found is not _exactly_ `6.0.0.0`. (Of course, which one of the two listed above it actually chooses -- `10586.0` or `10586.839` -- is open to speculation.) Hope this helps!

[the remark this answers is not in the document]
Comment (yagoulas) [reply boot/bootdata/packages/reactos.dff.in]: @learn-more : Every time ms changes the way the hash is computed (the last part of the name), the file names should change as well.

File: dll/ntdll/ldr/ldrpe.c [hunk 4]
```diff
@@ -964,6 +949,11 @@ LdrpLoadImportModule(IN PWSTR DllPath OPTIONAL,
                        &(*DataTableEntry)->InInitializationOrderLinks);
     }
 
+done:
+
+    /* Do we have a redirect string? */
+    if (DllString1.Buffer) RtlFreeUnicodeString(&DllString1);
```
[the remark this answers is not in the document]
Comment (yagoulas) [reply dll/ntdll/ldr/ldrpe.c]: Honestly I copied that part from another place in LDR which used RtlDosApplyFileIsolationRedirection_Ustr

File: dll/ntdll/ldr/ldrutils.c [hunk 5]
```diff
@@ -214,27 +119,38 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
+    BOOLEAN RedirectedDll;
+    RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED ActCtx;
+
+    /* Set up the Act Ctx */
+    ActCtx.Size = sizeof(ActCtx);
+    ActCtx.Format = RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_FORMAT_WHISTLER;
+    RtlZeroMemory(&ActCtx.Frame, sizeof(RTL_ACTIVATION_CONTEXT_STACK_FRAME));
+
+    /* Activate the ActCtx */
+    RtlActivateActivationContextUnsafeFast(&ActCtx,
```
[the remark this answers is not in the document]
Comment (yagoulas) [reply dll/ntdll/ldr/ldrutils.c]: Return value doesn't really indicate failure or success.

File: dll/ntdll/rtl/libsupp.c (hunk 1 above)
Comment (yagoulas) [reply dll/ntdll/rtl/libsupp.c]: would a boolean called bAllocate be better?

File: dll/ntdll/ldr/ldrutils.c (hunk 2 above)
Comment (yagoulas) [reply dll/ntdll/ldr/ldrutils.c]: I refrained from doing any kind of refactoring as this code was already duplicated.

File: dll/ntdll/ldr/ldrutils.c [hunk 6]
```diff
@@ -214,27 +119,38 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
     ULONG ImportSize;
     NTSTATUS Status;
     ULONG i;
+    BOOLEAN RedirectedDll;
+    RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED ActCtx;
+
+    /* Set up the Act Ctx */
+    ActCtx.Size = sizeof(ActCtx);
+    ActCtx.Format = RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_FORMAT_WHISTLER;
+    RtlZeroMemory(&ActCtx.Frame, sizeof(RTL_ACTIVATION_CONTEXT_STACK_FRAME));
```
[the remark this answers is not in the document]
Comment (yagoulas) [reply dll/ntdll/ldr/ldrutils.c]: Right, I copied this part from elsewhere as well but it doesn't really look correct.

File: dll/ntdll/rtl/libsupp.c (hunk 1 above)
Comment (learn-more) [reply dll/ntdll/rtl/libsupp.c]: imo, yes :)

File: boot/bootdata/packages/reactos.dff.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (spetreolle) [reply boot/bootdata/packages/reactos.dff.in]: This DLL shouldn't be installed as a regular file.
Windows has links to avoid using hundreds of identical files, and we already support them.

[the remark this answers is not in the document]
Comment (sdever) [reply boot/bootdata/packages/reactos.dff.in]: @spetreolle Afaik FAT32 doesn't support links (and ReactOS can't be installed on anything else than FAT32)

**Comment by jeditobe** [comment]
Any news with this? I believe it may fix CORE-9312 and other LDR-related bugs.

**Comment by yagoulas** [comment]
It will be comitted after the upcoming release as it can break a lot of stuff.

**Comment by encodedpr** [comment]
Status?

**Comment by yagoulas** [comment]
The last tests will be done soon and if everything is fine it will be merged. These are the test results:
https://www.reactos.org/testman/compare.php?ids=56588,56591
https://www.reactos.org/testman/compare.php?ids=56589,56592
https://www.reactos.org/testman/compare.php?ids=56590,56587

**Comment by spetreolle** [comment]
I see new failures in ntdll:RtlDosApplyFileIsolationRedirection_Ustr.
They weren't there in your previous patches.

**Comment by yagoulas** [comment]
That's because this PR doesn't fix all possible scenarios involving RtlDosApplyFileIsolationRedirection_Ustr. We don't have gdiplus in winsxs which causes failures, the test hardcodes c:\\windows\\ which also causes failures and it also covers .local redirection which isn't implemented and also causes failures. Other than that there are still some edge cases for this function that still fail.

File: dll/ntdll/ldr/ldrpe.c [hunk 7]
```diff
@@ -883,76 +887,57 @@ LdrpLoadImportModule(IN PWSTR DllPath OPTIONAL,
...
+    /* Check success */
+    if (NT_SUCCESS(Status))
+    {
+        /* Let Ldrp know */
+        RedirectedDll = TRUE;
+    }
+    else if (Status != STATUS_SXS_KEY_NOT_FOUND)
+    {
+        /* Unrecoverable SxS failure */
+        goto done;
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrpe.c]: A DPRINT would help debugging here.

File: dll/ntdll/ldr/ldrpe.c [hunk 8]
```diff
@@ -964,6 +949,10 @@ LdrpLoadImportModule(IN PWSTR DllPath OPTIONAL,
                        &(*DataTableEntry)->InInitializationOrderLinks);
     }
 
+done:
+    /* Do we have a redirect string? */
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrpe.c]: Just drop the comment 🙂

File: dll/ntdll/ldr/ldrpe.c [hunk 9]
```diff
@@ -883,76 +887,57 @@ LdrpLoadImportModule(IN PWSTR DllPath OPTIONAL,
...
                         NULL,
                         TRUE,
-                        FALSE,
+                        RedirectedDll,
                         DataTableEntry);
-
-    if (!NT_SUCCESS(Status)) return Status;
+    if (!NT_SUCCESS(Status))
+    {
+        goto done;
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrpe.c]: Same, DPRINT can't hurt.

File: dll/ntdll/ldr/ldrutils.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (zefklop) [reply dll/ntdll/ldr/ldrutils.c]: The name of the fonction seems to indicate that some locking is necessary. Is the loader lock enough ?

File: dll/ntdll/ldr/ldrutils.c [hunk 10]
```diff
@@ -336,13 +236,105 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
                     }
+                    else if (RedirectedDll)
+                    {
+                        DPRINT1("LDR: LdrpCheckForLoadedDll failed for redirected dll %wZ\n", RedirectedImportName);
+                    }
+                }
+                else
+                {
+                    /* Unrecoverable SxS failure */
+                    DPRINT1("LDR: RtlDosApplyFileIsolationRedirection_Ustr failed for dll %wZ\n", ImportNameUnic);
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrutils.c]: DPRINT the status also ?

File: dll/ntdll/ldr/ldrutils.c [hunk 11]
```diff
@@ -336,13 +236,105 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
+                                    DPRINT1("LDR: Flags %lu  %wZ (%lx)\n", Flags, RedirectedImportName, Entry->LoadCount);
+                                }
+                            }
+
+                            /* Recurse into this entry */
+                            LdrpUpdateLoadCount3(Entry, Flags, UpdateString);
+                        }
+                        else if (RedirectedDll)
+                        {
+                            DPRINT1("LDR: LdrpCheckForLoadedDll failed for redirected dll %wZ\n", RedirectedImportName);
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrutils.c]: Ditto. Output the status also

File: dll/ntdll/ldr/ldrutils.c [hunk 12]
```diff
@@ -336,13 +236,105 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
...
+                        }
+                    }
+                    else
+                    {
+                        /* Unrecoverable SxS failure */
+                        DPRINT1("LDR: RtlDosApplyFileIsolationRedirection_Ustr failed for dll %wZ\n", ImportNameUnic);
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrutils.c]: Here too

File: dll/ntdll/ldr/ldrutils.c [hunk 13]
```diff
@@ -1977,8 +2000,10 @@ LdrpCheckForLoadedDll(IN PWSTR DllPath,
 
     /* Look in the hash table if flag was set */
 lookinhash:
-    if (Flag)
+    if (Flag  /* the second check is a hack */ && !RedirectedDll)
     {
+        /* FIXME: if we get redirected dll it means that we also get a full path so we need to find its filename for the hash lookup */
```
Comment (zefklop) [inline dll/ntdll/ldr/ldrutils.c]: I don't see how that's a problem. What does "flag" mean in this context ? From the code below, that looks like "not a full path"

Comment (yagoulas) [reply dll/ntdll/ldr/ldrutils.c]: The thing is that at this point we should do a fast lookup uing the hash, We can do that for redirected dlls as well . However the hash is computed only based on the filename so I was lazy making it find the filename when I wrote  this.

File: dll/ntdll/ldr/ldrutils.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (yagoulas) [reply dll/ntdll/ldr/ldrutils.c]: This function is Unssafe compared to RtlActivateActivationContext which uses a cookie to protect the activation context stack from out of order activations. In an out of order deactivation RtlDeactivateActivationContext fails and RtlDeactivateActivationContext just destroys the stack.

File: dll/ntdll/ldr/ldrpe.c (hunk 8 above)
Comment (yagoulas) [reply dll/ntdll/ldr/ldrpe.c]: right. I had a check to see if there is something in the string first but it wasn't needed and I forgot to remove the comment when I removed the check.

**Comment by jeditobe** [comment]
This thing is most interesting for me to have it in trunk.

**Comment by yagoulas** [comment]
I'm doing a last test run for this. Is there any issue that needs to be addressed before merging?

**Comment by yagoulas** [comment]
Latest test results:
https://www.reactos.org/testman/compare.php?ids=57291,57296
https://www.reactos.org/testman/compare.php?ids=57294,57297
